src/modules/global-messages/converters.ts

```diff
@@ -54,40 +9,31 @@ export const globalMessageConverter = {
     snapshot: QueryDocumentSnapshot<GlobalMessageType>,
   ): GlobalMessageType | undefined {
     const data = snapshot.data();
+    const potential = {
+      id: snapshot.id,
+      active: data.active,
+      title: data.title,
+      body: data.body,
+      type: data.type,
+      subtle: data.subtle,
+      context: data.context,
+      isDismissable: data.isDismissable ?? false,
+      startDate: firestoreTimestampToDate(data.startDate),
+      endDate: firestoreTimestampToDate(data.endDate),
+    };
 
-    if (!hasNecessaryGlobalMessageTypeFields(data)) {
-      return undefined;
-    }
-
-    const active = data.active;
-    const body = mapToLocalizedStringArray(data.body);
-    const title = mapToLocalizedStringArray(data.title);
-    const context = mapToContexts(data.context);
-    const type = mapToMessageType(data.type);
-    const subtle = data.subtle ?? false;
-    const isDismissable = data.isDismissable ?? false;
-    const startDate = data.startDate;
-    const endDate = data.endDate;
+    const validated = globalMessageTypeSchema.safeParse(potential);
 
-    if (!body || !context || !type) return;
+    if (validated.success) {
+      return validated.data;
+    }
 
-    return {
-      id: snapshot.id,
-      type,
-      subtle,
-      active,
-      context,
-      body,
-      title,
-      isDismissable,
-      startDate,
-      endDate,
-    };
+    return undefined;
   },
 };
 
-function hasNecessaryGlobalMessageTypeFields(data: any) {
-  return (
-    'active' in data && 'body' in data && 'context' in data && 'type' in data
-  );
+function firestoreTimestampToDate(timestamp: any) {
+  if (!timestamp) return;
+  if (!(timestamp instanceof Timestamp)) return;
+  return timestamp.toDate();
```

Comment:
Maybe not cleaner, but for the record you could have done this

```suggestion
  return timestamp?.toDate?.();
```

Reply:
Neat! I didn't know one could use optional chaining for function calls 👀